include/sys/fs/zfs.h

```diff
@@ -501,7 +501,8 @@ typedef enum {
 
 typedef enum {
 	ZFS_REDUNDANT_METADATA_ALL,
-	ZFS_REDUNDANT_METADATA_MOST
+	ZFS_REDUNDANT_METADATA_MOST,
+	ZFS_REDUNDANT_METADATA_NONE
```

Comment:
I have added both the `ZFS_REDUNDANT_METADATA_SOME` and `ZFS_REDUNDANT_METADATA_NONE` options.
The behavior of some/none values w.r.t OTs:
`ZFS_REDUNDANT_METADATA_NONE` -> critical_metadata (`DMU_OT_OBJSET, DMU_OT_DSL_DIR, DMU_OT_DSL_PROPS, DMU_OT_DSL_DATASET, DMU_OT_MASTER_NODE, DMU_OT_ZVOL_PROP, DMU_OT_POOL_PROPS, DMU_OT_SA, DMU_OT_SA_MASTER_NODE, DMU_OT_SA_ATTR_REGISTRATION, DMU_OT_SA_ATTR_LAYOUTS`)

`ZFS_REDUNDANT_METADATA_SOME` ->`DMU_OT_IS_METADATA(type)`

Reply:
The MOS always uses ZFS_REDUNDANT_METADATA_ALL, so it isn't relevant to talk about the value of DMU_OT_IS_CRITICAL() for MOS object types (DMU_OT_DSL_DATASET, etc).  So the difference between SOME and NONE is primarily that SOME includes DMU_OT_DIRECTORY_CONTENTS whereas NONE does not?

I think what's implemented here is:
* NONE: essentially no extra copies in this dataset (of course the pool-wide metadata is still stored 3x, and some minor bits like SA stuff).  Any single block damage can take out the entire dataset.
* SOME: extra copy of metadata-type objects (including indirect blocks and data blocks) (e.g. dnodes, directories).  Any single block damage can take out at most one user file (or zvol).
* MOST: additionally, extra copy of L2 and above indirect blocks of user-type objects (e.g. files, zvols).  Any single block damage can take out at most 1024 logically-contiguous blocks of user data (e.g. 8MB, with recordsize=8k).
* ALL (default): additionally, extra copy of L1 indirect blocks of user-type objects.  Any single block damaged can take out at most 1 block of user data (e.g. 128K, with recordsize=128k).

It seems to me that distinguishing "critical" object types might not be necessary, if we are OK with NONE not having extra copies of the SA and other small bits of metadata.  It's true that they are critical for operating on the dataset, but you could say the same thing about the single L5 indirect block of a zvol, or the block of dnodes that contains the root directory.  So with NONE, any corruption in the dataset has a chance of losing the whole dataset, regardless of the "critical" object types.

I think it might be hard for admins to chose among these 4 options.  Maybe documenting (in the manpage) the "any single block damage can take out..." for each of the options, would help them understand the trade-off.  We currently do that for ALL and MOST, although in a slightly hand-wavey way, and also it says 100 where it should say 1000 (due to the long-ago indirect block size increase from 16KB -> 128KB).